src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -598,20 +598,39 @@ QString QgsPostgresUtils::whereClause( const QgsFeatureIds &featureIds, const Qg
 
     //simple primary key, so prefer to use an "IN (...)" query. These are much faster then multiple chained ...OR... clauses
     QString delim;
-    QString expr = QStringLiteral( "%1 IN (" ).arg( QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[0] ).name() ) );
+    QString whereKeys;
+    for ( int i = 0; i < pkAttrs.size(); i++ )
+    {
+      whereKeys += delim + QgsPostgresConn::quotedIdentifier( fields.at( pkAttrs[i] ).name() );
+      delim = QStringLiteral( "," );
+    }
 
+    QStringList whereValuesList;
     for ( const QgsFeatureId featureId : std::as_const( featureIds ) )
     {
       const QVariantList pkVals = sharedData->lookupKey( featureId );
       if ( !pkVals.isEmpty() )
       {
-        expr += delim + QgsPostgresConn::quotedValue( pkVals.at( 0 ) );
-        delim = ',';
+        Q_ASSERT( pkVals.size() == pkAttrs.size() );
+
+        QString delim;
+        QString whereValues;
+        for ( int i = 0; i < pkAttrs.size(); i++ )
+        {
+          whereValues += delim + QgsPostgresConn::quotedValue( pkVals[i] );
+          delim = QStringLiteral( "," );
+        }
+        whereValuesList << whereValues;
       }
     }
-    expr += ')';
+    if ( 1 ==  pkAttrs.size() )
```

Comment:
Could you just isolate the modification of whereKeys above the for loop on featureIds so we just add whereValuesList here.

And I thought it could be better to not make a special case of multi column pks and to always surround *whereKeys* with parenthesis (just to make the code a little bit simple).